README
2Performant Ad Groups
------------------------

This Wordpress plugin enables you to manage and show the ad groups in a widget on your blog.

api/Changelog
Changelog
=========

* Tue Jul 12 14:56:03 EEST 2011 - Tudor Sandu <tudor.sandu@2parale.ro>
  - 0.4.5 release
  - added TPException class and descendants and thrown them in appropriate locations, in case of errors
  - tweaked include path to make sure appropriate PEAR classes are loaded
  - renamed 2performant.php to TPerformant.php

* Wed Jun 15 12:30:58 EEST 2011 - Radu Spineanu <radu@2performant.com>
  - 0.4.4 release
  - added stats_ functions

* Tue May  3 20:23:58 EEST 2011 - Radu Spineanu <radu@2performant.com>
  - 0.4.3 release
  - added admin_{affiliate,advertiser}_commissions_list which
    return commission details

* Wed Apr 20 16:58:08 EEST 2011 - Radu Spineanu <radu@2performant.com>
  - 0.4.2 release
  - fixed admin request buiding to properly add host
  - disable SSL certificate checks for https calls

* Tue Apr 19 12:54:25 EEST 2011 - Radu Spineanu <radu@2performant.com>
  - 0.4.1 release
  - fixed admin_ methods to properly create requests
  - OAuth: build signature using encoded params in body

* Sat Apr 16 19:49:36 EEST 2011 - Radu Spineanu <radu@2performant.com>
  - 0.4 release
  - API through JSON

* Thu Feb 17 00:48:52 EET 2011 - Radu Spineanu <radu@2performant.com>
  - 0.3.1 release
  - add user_create method

* Sun Nov 28 16:09:54 EET 2010 - Radu Spineanu <radu@2performant.com>
  - 0.3.0 release
  - admin_api methods
  - changed from _merchant methods to _advertiser methods

* Wed Jun  9 01:36:52 EEST 2010 - Radu Spineanu <radu@2performant.com>
  - 0.2.5 release
  - adds hooks_* methods

* Thu May 13 19:29:33 EEST 2010 - Radu Spineanu <radu@2performant.com>
  - 0.2.4 release
  - set Content-Type to application/xml when using OAuth
  - disable default content-type in OAuth library
  - new methods: sale_create, lead_create, banner_create
    ad_group_createitem, feeds methods

* Tue Apr 20 14:26:16 EEST 2010 - Radu Spineanu <radu@2performant.com>
  - 0.2.3 release
  - changed initializer from &new to new

* Fri Apr  9 10:43:21 EEST 2010 - Radu Spineanu <radu@2performant.com>
  - 0.2.2 release
  - change the ordering of arguments in commissions->search()
    to avoid multiple null arguments in frequent use cases.
  - updated the readme with proper initialization call for simple.

* Wed Mar  3 20:01:03 PST 2010 - Radu Spineanu <radu@2performant.com>
  - 0.2.1 release
  - changed methods from widget_store to product_store

* Mon Feb  8 16:44:05 PST 2010 - Radu Spineanu <radu@2performant.com>
  - 0.2 release
  - Added PEAR libraries to the tarball.
  - API Versioning.
  - OAuth support.
  - A couple of new user_ methods.
  - Fewer arguments in commissions_search (uses hash).

* Mon Apr 23 12:01:49 EET 2009 - Radu Spineanu <radu@2performant.com>
  - 0.1.6 release
  - Added XML_SERIALIZER_OPTION_CDATA_SECTIONS to serialize().
    Thanks to Valentin Agachi for the patch.
  - Added a new option to _search called sort.
  - Added a new option to WidgetStores list/search called uniq_products.
    It makes sure the campaign prid pair is unique so it only returns 
    uniq products.

* Fri Feb 27 19:58:50 EET 2009 - Radu Spineanu <radu@2performant.com>
  - 0.1.5 release
  - Fixed minor bug where widget_store was expected instead of widget-store.

* Wed Feb 18 14:02:37 EET 2009 - Radu Spineanu <radu@2performant.com>
  - 0.1.4 release
  - Added possibility of searching after transaction_ids in commissions.
  - commissions_listformerchant() should expect campaign objects.
  - fixed commissions_update() so it actually works.

* Wed Jan 28 14:50:20 EET 2009 - Radu Spineanu <radu@2performant.com>
  - 0.1.3 release
  - changed the widget_store_search_products() arguments since it doesn't
    need a widget_store_id anymore

* Mon Jan 12 00:03:38 EET 2009 - Radu Spineanu <radu@2performant.com>
  - 0.1.2 release
  - renamed widget_store_showproducts() to widget_store_showitems() for code
    unity.
  - widget_store_showitems() should expect widget-store-data
  - added widget_store_showitem() that's used when we know a product's id
    (NOT prid)

* Tue Dec 23 17:34:48 EET 2008 - Radu Spineanu <radu@2performant.com>
  - 0.1.1 release
  - send $request with txtlinks_list(), txtads_list() and $banner_list()
    resulting in working pagination

* Mon Nov 10 10:33:54 EET 2008 - Radu Spineanu <radu@2performant.com>
  - 0.1 released
  - initial release


widget.php
<?php

/* TPAG WIDGET */
class TPAG_Widget extends WP_Widget {
	
	// the widget constructor
    function TPAG_Widget() {
        parent::WP_Widget(false, $name = '2Performant Ad Groups');
		parent::__construct(__CLASS__, '2Performant Ad Groups', array(
			'classname' => __CLASS__,
			'description' => "This widget will show ad groups from 2Performant platform."
		));
    }


	// content of the widget - front-end
    function widget($args, $instance) {
		extract( $args );
		$title     = apply_filters('widget_title', $instance['title']);
		?>
			<?php echo $before_widget; ?>
			<?php if ( $title )
				echo $before_title . $title . $after_title; ?>

			<script src='http://event.2parale.ro/ad_group/<?php echo $instance['group_name']; ?>_<?php echo $instance['affiliate_id']; ?>.js' type='text/javascript'></script>

			<?php echo $after_widget; ?>
		<?php
    }

	
	// get the options
    function update($new_instance, $old_instance) {
		$instance = $old_instance;
		$instance['group_name']     = strip_tags($new_instance['group_name']);
		$instance['affiliate_id']   = strip_tags($new_instance['affiliate_id']);
		return $instance;
    }

	
	// the form that shows in your administration area - here you insert the widget options
    function form($instance) {
		$tp = new TPerformant_Wrapper;
		$tp->connection();

		$ad_groups_list = $tp->api->ad_groups_list();
		$user_details   = $tp->api->user_loggedin();
        $group_name  = esc_attr($instance['group_name']);
		?>
		<p>
			<label for="<?php echo $this->get_field_id('group_name'); ?>">Group name:</label>
			<select id="<?php echo $this->get_field_id('group_name'); ?>" name="<?php echo $this->get_field_name('group_name'); ?>">
				<?php
				if( !empty($ad_groups_list) ) {print_r($ad_groups_list);
					foreach($ad_groups_list->{'ad-group'} as $ad) {
						echo '<option value="'.$ad->{'unique-code'}.'" '.($group_name == $ad->{'unique-code'} ? 'selected="selected"' : '').'>'.$ad->name.'</option>';
					} 
				}
				?>
			</select>
		</p>
		<input type="hidden" name="<?php echo $this->get_field_name('affiliate_id'); ?>" value="<?php echo $user_details->{'unique-code'}; ?>" />
        <?php 
    }
   
}


api/2performant_wrapper.php
<?php

class TPerformant_Wrapper {

	public $api;

	// 2perfomant wrapper
	public function wrapper() {
		if ( !isset( $this->api ) ) {
			$err = $this->connection();
			if( !empty($err) )
				return false;
		}
	}
	
	
	// create a connection with 2Performant API
	public function connection() {
		$connected = false;
		$errors = array();
		$connection = get_option( 'tpag_options_connection' );
		if ( ! ( $connection['network'] && $connection['username'] && $connection['password'] ) ) {
			$errors[] = 'Please define connection parameters';
		} else {
			try {
				$api_options = array_merge( array( 'connection_timeout' => 10, 'timeout' => 0, 'adapter' => 'curl' ), $connection );
				$config = array( 'HTTP_Request2_config' => array(
					'connect_timeout' => $api_options['connection_timeout'],
					'timeout' => $api_options['timeout'],
					'adapter' => 'HTTP_Request2_Adapter_' . ucfirst($api_options['adapter'])
				) );
				$this->api = new TPerformant( 'simple', array( "user" => trim( $connection['username'] ), "pass" => trim( $connection['password'] ) ), trim( $connection['network'] ), $config );
				
				try {
					$response = $this->api->user_loggedin();
				} catch( Exception $e ) {
					$error = $e->getMessage();
					function starts_with( $haystack, $needle ) { return substr( $haystack, 0, strlen( $needle ) ) == $needle; }
					if ( starts_with( $error, 'Unavailable server. Response code: 401' ) ) {
						$errors[] = 'Invalid username and password';
					} elseif ( starts_with( $error, 'Unavailable server. Response code: ' ) ) {
						$errors[] = 'Invalid API URL';
					}
				}
				if( isset( $response ) && $response ) {
					if ( is_object( $response ) ) {
						$connected = true;
						
						if ( ! ( isset( $response->role ) && $response->role == 'affiliate' ) )
							$errors[] = 'Logged in user is not an affiliate';
					} else {
						$errors[] = 'Invalid API response. Please verify settings and contact affiliate network administrator.';
						ob_start();
						//print_r( $response );
						trigger_error( 'API response: ' . ob_get_contents(), E_USER_WARNING );
						ob_end_clean();
					}
				}
			} catch( HTTP_Request2_Exception $e ) {
				$errors[] = 'API connection error: ' . $e->getMessage();
			} catch( Exception $e ) {
				$errors[] = get_class($e) . ': ' . $e->getMessage();
			}
		}
		return $errors;
	}

}

plugin.php
<?php
/*
Plugin Name: 2Performant Ad Groups
Plugin URI: http://siteafiliere.com
Description: This plugin allows you to manage your ad groups from 2Performant platform and show the ads in your website.
Author: Alexandru Raduta
Version: 0.1
Author URI: http://siteafiliere.com
*/

$pear_dir = WP_PLUGIN_DIR . '/' . str_replace( basename( __FILE__ ), '', plugin_basename( __FILE__ ) ) . 'api';
set_include_path( get_include_path() . PATH_SEPARATOR . $pear_dir . PATH_SEPARATOR . $pear_dir . '/PEAR' );
include_once 'api/2performant.php';
include_once 'api/2performant_wrapper.php';
include_once 'widget.php';


class TPerformantAdGroups {

	public $tp;

	public function __construct() {
		// add a menu for the settings page
		add_action('admin_menu', array(&$this, 'tpag_settings_menu'));
		// register plugin settings
		add_action('admin_init', array(&$this, 'tpag_register_plugin_settings'));
		// register tpag widget
		add_action('widgets_init', create_function('', 'register_widget("TPAG_Widget");'));
	}
	
	
	// load template file
	public function render($name, $vars = array()) {
		if( isset($vars) && is_array($vars) && !empty($vars) ) {
			extract($vars);
		}
		// require template file
		require 'views/' . $name . '.php';
	}
	
	
	// add a menu
	public function tpag_settings_menu() {
		add_options_page( '2Performant Ad Groups Settings', '2Performant Ad Groups Settings', 'manage_options', '2performant-ad-groups', array(&$this, 'tpag_settings_page') );
	}
	
	
	// settings page
	public function tpag_settings_page() {
		// verify user permissions
		if ( ! current_user_can( 'manage_options' ) )  {
			wp_die( __( 'You do not have sufficient permissions to access this page.' ) );
		}

		// if the user entered the connections settings, verify if he can connect to the 2Perfomant API url
		$connection = get_option( 'tpag_options_connection' );
		$tp = new TPerformant_Wrapper();
		if( !empty($connection) ) { 
			$data['errors'] = $tp->connection();
		} else {
			$data['errors'] = array();
		}
		
		// render template
		$this->render('settings-page', $data);
	}
	
	
	// register plugin settings
	public function tpag_register_plugin_settings() {
		$settings = $this->tpag_settings_fields_array();
		if( !empty($settings) ) {
			foreach ( $settings as $section_id => $section ) {
				register_setting( 'tpag-options-group', "tpag_options_{$section_id}" );
				add_settings_section( "tpag_options_{$section_id}", $section['label'], array(&$this, 'tpag_options_callback'), '2performant-ad-groups' );
				// get saved settings
				$values = get_option( "tpag_options_{$section_id}" );

				foreach ( $section['settings'] as $setting_id => $setting ) {
					$setting['id'] = "tpag_options_{$section_id}_{$setting_id}";
					$setting['name'] = "tpag_options_{$section_id}[{$setting_id}]";
					$setting['value'] = $values[$setting_id];
					add_settings_field( "tpag_options_{$section_id}_{$setting_id}", $setting['label'], array(&$this, 'tpag_render_settings_fields'), '2performant-ad-groups', "tpag_options_{$section_id}", $setting );	
					register_setting('2performant-ad-groups', "tpag_options_{$section_id}_{$setting_id}");
				}
			}
		}
	}
	
	
	// fields for the settings page
	public function tpag_settings_fields_array($section = false, $setting = false) {
		$settings = array(
			'connection' => array(
				'label' => 'Connection settings for the affiliate network',
				'settings' => array(
					'network' => array(
						'type' => 'text',
						'label' => 'Network API URL',
						'description' => 'E.g. http://api.2parale.ro'
					),
					'username' => array(
						'type' => 'text',
						'label' => 'Username',
						'description' => 'E.g. your username from http://2parale.ro'
					),
					'password' => array(
						'type' => 'password',
						'label' => 'Password',
						'description' => 'E.g. your password from http://2parale.ro'
					)
				)
			)
		);
		
		if ( $section !== false ) 
		if ( isset( $settings[$section] ) ) {
			$settings = $settings[$section];

			if( $setting !== false )
			if ( isset( $settings['settings'][$setting] ) ) {
				$settings = $settings['settings'][$setting];
			} else {
				return false;
			}
		} else {
			return false;
		}
		
		return $settings;
	}
	
	
	// first heading on settings page
	public function tpag_options_callback() {
		//
	}
	
	
	// render settings page fields
	public function tpag_render_settings_fields($setting) {
		extract( $setting );
		$type = empty( $type ) ? 'text' : $type;
		$class = isset( $class ) ? $class : array();
		$class = is_array( $class ) ? $class : array ( $class );

		$output = '';
		switch ( $type ) {
			case 'select':
				if ( isset( $options ) && is_array( $options ) ) {
					foreach ( $options as $key => $option ) {
						$selected = $key == $value ? 'selected="selected"' : '';
						$output .= "<option value='".esc_attr($key)."' {$selected}>" . esc_attr( $option ) . "</option>";
					}
				}
				$class = implode( ' ', $class );
				$output = "<select name='" . esc_attr( $name ) . "' id='" . esc_attr( $id ) . "' class='".esc_attr( $class ) . "'>$output</select>";
				break;
			case 'text':
			case 'number':
			case 'password':
				$class = array_merge( $type == 'number' ? array( 'small-text' ) : array ( 'regular-text' ), $class );
				$class = implode( ' ', $class );
				$output = "<input type='" . esc_attr( $type == 'number' ? 'text' : $type ) . "' class='" . esc_attr( $class ) . "' name='" . esc_attr( $name ) . "' id='" . esc_attr( $id ) . "' value='" . esc_attr( $value ) . "' />";
				break;
			case 'hidden':
				$output = "<input type='hidden' name='" . esc_attr( $name ) . "' id='" . esc_attr( $id ) . "' value='" . esc_attr( $value ) . "' />";
				break;
			case 'textarea':
				$class = array_merge( array ( 'large-text', 'code' ), $class );
				$class = implode( ' ', $class );
				$output = "<textarea rows='10' cols='50' class='" . esc_attr( $class ) . "' name='" . esc_attr( $name ) . "' id='" . esc_attr( $id ) . "'>" . esc_attr( $value ) . "</textarea>";
				break;
			case 'checkbox':
				$class = implode( ' ', $class );
				$output = "<input type='checkbox' class='" . esc_attr( $class ) . "' name='" . esc_attr( $name ) . "' id='" . esc_attr( $id ) . "' ".((boolean)($value) ? "checked='checked'" : "")." />";
				break;
			case 'custom':
				if ( isset( $callback ) ) {
					$output .= call_user_func( $callback, $setting );
				}
				break;
			default:
				trigger_error( 'Invalid setting type', E_WARNING );
		}

		if ( isset( $description ) )
			$output .= " <span class='description'>" . esc_attr( $description ) . "</span>";

		echo $output;
	}
	
}

new TPerformantAdGroups();


